core/finality-grandpa/src/finality_proof.rs

```diff
@@ -17,254 +17,605 @@
+	pub headers_to_import: Vec<Header>,
+	/// The hash of the block that could be finalized.
+	pub block: Header::Hash,
+	/// The justification for the block.
+	pub justification: Vec<u8>,
+	/// New authorities set id that should be applied starting from block.
+	pub new_set_id: u64,
+	/// New authorities set that should be applied starting from block.
+	pub new_authorities: Vec<(AuthorityId, u64)>,
+}
+
+/// Single fragment of proof-of-finality.
+///
+/// Finality for block B is proved by providing:
+/// 1) the justification for the descendant block F;
+/// 2) headers sub-chain (B; F] if B != F;
+/// 3) proof of GRANDPA::authorities() if the set changes at block F.
+#[derive(Debug, PartialEq, Encode, Decode)]
+struct FinalityProofFragment<Header: HeaderT> {
+	/// The hash of block F for which justification is provided.
+	pub block: Header::Hash,
+	/// Justification of the block F.
+	pub justification: Vec<u8>,
+	/// The set of headers in the range (U; F] that we believe are unknown to the caller. Ordered.
+	pub unknown_headers: Vec<Header>,
+	/// Optional proof of execution of GRANDPA::authorities().
+	pub authorities_proof: Option<Vec<Vec<u8>>>,
+}
+
+/// Proof of finality is the ordered set of finality fragments, where:
+/// - last fragment provides justification for the best possible block from the requested range;
+/// - all other fragments provide justifications for GRANDPA authorities set changes within requested range.
+type FinalityProof<Header> = Vec<FinalityProofFragment<Header>>;
+
+/// Finality proof request data.
+#[derive(Debug, Encode, Decode)]
+enum FinalityProofRequest<H: Encode + Decode> {
+	/// Original version of the request.
+	Original(OriginalFinalityProofRequest<H>),
+}
+
+/// Original version of finality proof request.
+#[derive(Debug, Encode, Decode)]
+struct OriginalFinalityProofRequest<H: Encode + Decode> {
+	/// The authorities set id we are waiting proof from.
+	///
+	/// The first justification in the proof must be signed by this authority set.
+	pub authorities_set_id: u64,
+	/// Hash of the last known finalized block.
+	pub last_finalized: H,
+}
+
+/// Prepare data blob associated with finality proof request.
+pub(crate) fn make_finality_proof_request<H: Encode + Decode>(last_finalized: H, authorities_set_id: u64) -> Vec<u8> {
+	FinalityProofRequest::Original(OriginalFinalityProofRequest {
+		authorities_set_id,
+		last_finalized,
+	}).encode()
+}
+
+/// Prepare proof-of-finality for the best possible block in the range: (begin; end].
 ///
-/// The proof is the serialized `FinalityProof` constructed using earliest known
-/// justification of the block. None is returned if there's no known justification atm.
-pub fn prove_finality<Block: BlockT, B, G>(
+/// It is assumed that the caller already have a proof-of-finality for the block 'begin'.
+/// It is assumed that the caller already knows all blocks in the range (begin; end].
+///
+/// Returns None if there are no finalized blocks unknown to the caller.
```

Comment:
You're right, that's a good point. Thanks! I'l think of how we could handle it. In practice, however, it shouldn't be that big, because:
1) if there's major syncing - we fetch justifications for each block && if we're unable to verify justification (i.e. GRANDPA authorities set has changed) => we will request finality proof && it will be of length = 1;
2) if we're following the tip of the chain, then either we will receive block with ready justification (i.e. (1) works here) or, if justification isn't ready yet, I could remember that we had a plans (not sure that is done atm) to announce it once it is ready (this should be handled by the light client, though).

So the only case when `len(proof) > 1` is when we have missed several justifications announcements.

Reply:
Limited to 8 fragments now, because every fragment includes a proof-of-execution && it is rather large atm. But might change this criteria later (e.g. to limit by number of bytes instead of simply fragments count).